crates/ide-assists/src/utils.rs

```diff
@@ -316,43 +316,52 @@ pub(crate) fn does_pat_match_variant(pat: &ast::Pat, var: &ast::Pat) -> bool {
     pat_head == var_head
 }
 
-pub(crate) fn does_nested_pattern(pat: &ast::Pat) -> bool {
-    let depth = calc_depth(pat, 0);
+pub(crate) fn does_pat_variant_nested_or_literal(pat: &ast::Pat) -> bool {
+    check_pat_variant_nested_or_literal_with_depth(pat, 0)
+}
 
-    if 1 < depth {
+fn check_pat_variant_nested_or_literal_with_depth(pat: &ast::Pat, depth: usize) -> bool {
+    if depth > 1 {
         return true;
     }
-    false
-}
 
-fn calc_depth(pat: &ast::Pat, depth: usize) -> usize {
     match pat {
         ast::Pat::IdentPat(_)
-        | ast::Pat::BoxPat(_)
         | ast::Pat::RestPat(_)
-        | ast::Pat::LiteralPat(_)
-        | ast::Pat::MacroPat(_)
-        | ast::Pat::OrPat(_)
-        | ast::Pat::ParenPat(_)
-        | ast::Pat::PathPat(_)
         | ast::Pat::WildcardPat(_)
-        | ast::Pat::RangePat(_)
-        | ast::Pat::RecordPat(_)
         | ast::Pat::RefPat(_)
-        | ast::Pat::SlicePat(_)
-        | ast::Pat::TuplePat(_)
-        | ast::Pat::ConstBlockPat(_) => depth,
+        | ast::Pat::BoxPat(_)
+        | ast::Pat::ConstBlockPat(_) => false,
+
+        ast::Pat::LiteralPat(_)
+        | ast::Pat::RangePat(_)
+        | ast::Pat::MacroPat(_)
+        | ast::Pat::PathPat(_) => true,
+
+        ast::Pat::ParenPat(pat) => pat
+            .pat()
+            .map_or(true, |pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth)),
+        ast::Pat::TuplePat(pat) => {
+            pat.fields().any(|pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth))
```

Comment:
```suggestion
            pat.fields().any(|pat| check_pat_variant_nested_or_literal_with_depth(&pat, depth + 1))
```

Should be `depth + 1`.

Reply:
I think nesting tuple pattern is fine if it does not contain literal patterns or tuple struct patterns or else.

```rust
let bar: Result<(i32, i32, Option<i32>), ()> = Ok((1, 2, Some(3)));
if let Ok((first, second, third)) = bar {} // fine
if let Ok((1, second, third)) = bar {} // bad
if let Ok((first, second, Some(third))) = bar {} // bad
```

Reply:
Then why the others (e.g. `OrPat`) have a depth limit?

Reply:
Fair point, I must have missed it. Just fixed it.

Reply:
It's also in `TupleStructPat`, and if you're going to not use `depth`, why not just remove it?

Reply:
I doubt about `TupleStructPat`, it should not be nested.
```rust
let bar: Result<Option<i32>, ()> = Ok(Some(1));
if let Ok(Some(a)) = bar {} // bad
```

Reply:
Ah so this is the reason - then `RecordPat` should also be `depth + 1` (`Enum::Variant {}`) - and perhaps also rename `depth` to something more informative, e.g. `depth_after_refutable`.

Reply:
Fair. You remind me that I forgot to consider record in enum `Enum::Variant {}` at all. My current test cases only covered structs `MyStruct { a, b }`.

Reply:
Fixed. I also covered cases like `MyStruct(i32, i32)`